### PR TITLE
Add file validation, show errors and allow user to add more files

### PR DESCRIPTION
**Pull request by romifz**
ref #43

### CODE REVIEW

**Comment by jpmckinney** [comment]
Can you fix merge conflicts? I can then review.

**Comment by romifz** [comment]
@jpmckinney merge conflicts are fixed.

**Comment by jpmckinney** [comment]
@romifz Strange – did you push? GitHub still reports conflicts needing to be resolved.

**Comment by romifz** [comment]
It's because of the last merge I think, I'll check the conflicts.

**Comment by romifz** [comment]
Now it's done.

File: default/views.py [hunk 1]
```diff
@@ -217,7 +224,31 @@ def uploadfile(request):
...
+        request.session['files'].remove(item)
+    request.session.modified = True
+
+    file_found = len(to_delete) > 0
+    response = JsonResponse({
+        'message': _('File deleted') if file_found else _('File not found')
+    })
+
+    if not file_found:
+        response.status_code = 404
```
Comment (jpmckinney) [inline default/views.py]: I find having a filter and then a loop a little less straightforward than it could be. The following seems easier to read:

```python
found = False
for fileinfo in request.session['files']:
    if fileinfo['id'] == str(id):
      request.session['files'].remove(item)
      found = True
```

I also think it'd be easier to read to put all 'file found' logic together, e.g.:

```python
if found:
    message = _('File deleted')
else:
    message = _('File not found')
    response.status_code = 404

response = JsonResponse({'message': message})
```

I'm assuming that the session can have multiple files with the same ID? Otherwise, if there can only be one, we can do:

```python
for fileinfo in request.session['files']:
    if fileinfo['id'] == str(id):
        request.session['files'].remove(item)
        message = _('File deleted')
        break
else:
    message = _('File not found')
    response.status_code = 404
```

File: default/data_file.py [hunk 2]
```diff
@@ -91,9 +99,30 @@ def write_json_to_zip(self, files):
         if isinstance(files, dict):
             files = files.items()
         for name, content in files:
-            with ZipFile(self.path, 'w', compression=ZIP_DEFLATED) as zipfile:
+            with ZipFile(self.path, 'a', compression=ZIP_DEFLATED) as zipfile:
                 zipfile.writestr(name, json_dumps(content) + '\n')
 
+    def is_valid(self, file, ocds_type=None):
```
[the remark this answers is not in the document]
Comment (jpmckinney) [reply default/data_file.py]: Nevermind the suggestion about an exception – I updated my comment to use use a boolean.

**Comment by jpmckinney** [comment]
@romifz Just checking that you weren't waiting for anything from me on this PR?

**Comment by romifz** [comment]
@jpmckinney sorry James, I hadn't had time to check this but I'll do this week. Thanks!

File: tests/__init__.py [hunk 3]
```diff
@@ -38,7 +38,7 @@ def upload_and_go(self, data=None):
     def assertResults(self, data, results):
         content = self.upload_and_go(data)
 
-        self.assertEqual(len(content), 2)
+        self.assertEqual(len(content), 3)
```
[the remark this answers is not in the document]
Comment (romifz) [reply tests/__init__.py]: I've changed the `_json_response` method to include `warnings` as a third key to be received in the client side. Now that I think about it, it should only appear when there are warnings, so I'll change this.

File: default/decorators.py [hunk 4]
```diff
@@ -35,8 +36,9 @@ def wrap(request, *args, **kwargs):
                 parser.parse(published_date)
                 kwargs['published_date'] = published_date
             except ValueError:
-                # TODO send a warning to client side
-                logger.debug('Invalid date submitted: {}, ignoring'.format(published_date))
+                msg = _('Invalid date submitted: {}, omitted from results.').format(published_date)
+                logger.debug(msg)
```
[the remark this answers is not in the document]
Comment (romifz) [reply default/decorators.py]: No, I don't think so.

File: default/views.py [hunk 5]
```diff
@@ -75,13 +76,14 @@ def _get_files_from_session(request):
         yield DataFile(**fileinfo)
 
 
-def _json_response(files):
+def _json_response(files, warnings=()):
```
[the remark this answers is not in the document]
Comment (romifz) [reply default/views.py]: ok! I'll change it.

File: default/views.py [hunk 6]
```diff
@@ -217,7 +224,31 @@ def uploadfile(request):
...
         }],
     })
+
+
+@require_GET
+def deletefile(request, id):
+    if 'files' not in request.session:
+        return JsonResponse({
+            'message': _('No files to delete')
+        }, status=404)
```
[the remark this answers is not in the document]
Comment (jpmckinney) [reply default/views.py]: Actually, 404 was right – I'll change it back :)

**Comment by jpmckinney** [comment]
@romifz I merged with master and added validation support to all views. You might want to check `invalid_request_file_message` in util.py (was `file_is_valid`).

I had to add more translateable strings. Can you update django.po?

```
#: default/util.py:74
msgid "Not a release or package"
msgstr ""

#: default/util.py:78
msgid "Not a package or list of packages"
msgstr ""

#: default/util.py:82
msgid "Not a release or list of releases"
msgstr ""

#: default/util.py:84
#, python-format
msgid "\"%(type)s\" not recognized"
msgstr ""
```

File: default/data_file.py [hunk 7]
```diff
@@ -91,7 +92,7 @@ def write_json_to_zip(self, files):
         if isinstance(files, dict):
             files = files.items()
         for name, content in files:
-            with ZipFile(self.path, 'w', compression=ZIP_DEFLATED) as zipfile:
+            with ZipFile(self.path, 'a', compression=ZIP_DEFLATED) as zipfile:
```
Comment (jpmckinney) [inline default/data_file.py]: Why was this changed to append mode?

File: default/templates/default/base-uploader.html [hunk 8]
```diff
@@ -15,8 +15,16 @@
     <b>{% trans "Success!" %}</b> result.zip(<span class="file-size"></span>)
     <a href="javascript:void(0)" class="download" target="_blank">{% trans "Download" %}</a>
 </div>
+<div class="response-warning action-failed alert alert-warning hidden">
+    <ul></ul>
+</div>
+<div class="response-warning file-process-failed alert alert-warning hidden">
+    {% trans "Some of the files could not be validated." %}
+    {% trans "You can add more files or remove them, then click on <b>Start</b> again." %}
+    {% trans "Invalid files will be ignored by default." %}
```
Comment (jpmckinney) [inline default/templates/default/base-uploader.html]: Can we be clearer on what we mean? e.g.

"We couldn't check that all files were OCDS JSON."
…
"Files that are not OCDS JSON will be ignored by default."

File: default/templates/default/to-spreadsheet.html [hunk 9]
```diff
@@ -110,13 +113,23 @@
...
                             $('.response-fail').removeClass('hidden');
                             })
                         .always(function(){
                             $('#processing-modal').modal('hide');
                             });
                 })
-                .always(function(){
+                .fail(function(jqXHR, textStatus, errorThrown){
+                    $('.response-fail').html(
+                        '{% trans "<b>Error in file uploaded</b>: " %}'
```
Comment (jpmckinney) [inline default/templates/default/to-spreadsheet.html]: Should this be "The file could not be uploaded"?

File: default/util.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+                    not isinstance(data, list) and not is_package(data)):
+                return _('Not a package or list of packages')
+        elif file_type == 'release release-array':
+            if (isinstance(data, list) and any(not is_release(item) for item in data) or
+                    not isinstance(data, list) and not is_release(data)):
+                return _('Not a release or list of releases')
+        else:
+            return _('"%(type)s" not recognized') % {'type': file_type}
+    except json.JSONDecodeError:
+        logger.debug('Error decoding file {}'.format(f.name))
```
Comment (jpmckinney) [inline default/util.py]: Do we need the logger? If not, delete the import and variable earlier in the file as well.

File: default/data_file.py (hunk 7 above)
Comment (romifz) [reply default/data_file.py]: Because the Upgrade feature expects to return all files in a single zip. When using `w` only the last file is returned.

File: default/templates/default/base-uploader.html (hunk 8 above)
Comment (romifz) [reply default/templates/default/base-uploader.html]: ok!

File: default/templates/default/to-spreadsheet.html (hunk 9 above)
Comment (romifz) [reply default/templates/default/to-spreadsheet.html]: It might be that the file wasn't uploaded, but it could be another validation error as well. I'll test and add more explicit messages here.

File: default/util.py (hunk 10 above)
Comment (romifz) [reply default/util.py]: I think it is ok for debug level?

Comment (jpmckinney) [reply default/util.py]: I'm just wondering if anyone reads these messages? This line is followed by `return _('Error decoding JSON')`, which ends up communicating the error to the frontend.

File: default/data_file.py (hunk 7 above)
Comment (jpmckinney) [reply default/data_file.py]: I think if we swap the position of the for-loop and with-statement, then we can use regular 'w' mode.

File: default/templates/default/base-uploader.html (hunk 8 above)
Comment (jpmckinney) [reply default/templates/default/base-uploader.html]: Also, I couldn't find where the logic is to ignore non-OCDS files?

File: default/data_file.py (hunk 7 above)
Comment (romifz) [reply default/data_file.py]: ok! I'll try.

File: default/templates/default/base-uploader.html (hunk 8 above)
Comment (romifz) [reply default/templates/default/base-uploader.html]: I guessed an user would like to exclude an invalid file (invalid means that it is not JSON, or not a release/record package in places a package is expected). Invalid JSON files and non-package files can crash the operations to make, that's why I think they should be ignored. We can remove this behaviour and let the user choose.

To determine a file is OCDS or not involves CoVE validation I think? Is this something we will like to do inside Toucan?

File: default/util.py (hunk 10 above)
Comment (romifz) [reply default/util.py]: I agree, I'll remove this line.

File: default/templates/default/base-uploader.html (hunk 8 above)
Comment (jpmckinney) [reply default/templates/default/base-uploader.html]: I think it's okay to ignore the files, as long as the user sees which files were ignored. Since this PR has been open a while, we can open a new issue to explore that.

OCDS Kit can work with data that doesn't pass CoVE validation, so I don't think we need to do that extra validation. I think checking for the presence of dict keys is enough for Toucan.

Comment (romifz) [reply default/templates/default/base-uploader.html]: Yep, I thought that just parsing and checking some keys would be fine. The user will see a message for each file that has an issue as in the image.

screenshot-127 0 0 1_8000-2019 12 19-15_13_28

Comment (jpmckinney) [reply default/templates/default/base-uploader.html]: Looks great!

**Comment by jpmckinney** [comment]
Thanks for your patience on all the review comments. I'll deploy the updates now!
